pvlib/iotools/srml.py

```diff
@@ -222,3 +228,85 @@ def read_srml_month_from_solardat(station, year, month, filetype='PO',
+
+
+def get_srml(station, start, end, filetype='PO', map_variables=True,
+             url="http://solardat.uoregon.edu/download/Archive/"):
+    """Request data from UoO SRML and read it into a Dataframe.
+
+    The Univeristy of Oregon Solar Radiation Monitoring Laboratory (SRML) is
+    described in [1]_. A list of stations can be found in [2]_.
+
+    Parameters
+    ----------
+    station : str
+        Two letter station abbreviation.
+    start : datetime like
+        First day of the requested period
+    end : datetime like
+        Last day of the requested period
+    filetype : string, default: 'PO'
+        SRML file type to gather. See notes for explanation.
+    map_variables : bool, default: True
+        When true, renames columns of the DataFrame to pvlib variable names
+        where applicable. See variable :const:`VARIABLE_MAP`.
+    url : str, default: 'http://solardat.uoregon.edu/download/Archive/'
+        API endpoint URL
+
+    Returns
+    -------
+    data : pd.DataFrame
+        Dataframe with data from SRML.
+    meta : dict
+        Metadata. Currently no metadata is parsed.
+
+    Notes
+    -----
+    File types designate the time interval of a file and if it contains
+    raw or processed data. For instance, `RO` designates raw, one minute
+    data and `PO` designates processed one minute data. The availability
+    of file types varies between sites. Below is a table of file types
+    and their time intervals. See [1] for site information.
+
+    ============= ============ ==================
+    time interval raw filetype processed filetype
+    ============= ============ ==================
+    1 minute      RO           PO
+    5 minute      RF           PF
+    15 minute     RQ           PQ
+    hourly        RH           PH
+    ============= ============ ==================
+
+    References
+    ----------
+    .. [1] University of Oregon Solar Radiation Measurement Laboratory
+       `http://solardat.uoregon.edu/ <http://solardat.uoregon.edu/>`_
+    .. [2] Station ID codes - Solar Radiation Measurement Laboratory
+       `http://solardat.uoregon.edu/StationIDCodes.html
+       <http://solardat.uoregon.edu/StationIDCodes.html>`_
+    """
+    # Use pd.to_datetime so that strings (e.g. '2021-01-01') are accepted
+    start = pd.to_datetime(start)
+    end = pd.to_datetime(end)
+
+    # Generate list of months
+    months = pd.date_range(
+        start, end.replace(day=1) + pd.DateOffset(months=1), freq='1M')
+    months_str = months.strftime('%y%m')
+
+    # Generate list of filenames
+    filenames = [f"{station}{filetype}{m}.txt" for m in months_str]
+
+    dfs = []  # Initialize list of monthly dataframes
+    for f in filenames:
+        try:
+            dfi = read_srml(url + f, map_variables=map_variables)
+            dfs.append(dfi)
+        except urllib.error.HTTPError:
+            warnings.warn(f"The following file was not found: {f}")
+
+    data = pd.concat(dfs, axis='rows')
```

Comment:
I am not opposed to this suggestion, although we may run into some issues related to timezone. I think this is why the `start`/`end` parameters in Solar Forecast Arbiter are required to be timezone localized - that seems like a hassle though.

There already exist several functions in pvlib that return a full month when requesting a single day btw, e.g. `get_bsrn`, and probably others too.

Reply:
> I am not opposed to this suggestion, although we may run into some issues related to timezone. I think this is why the start/end parameters in Solar Forecast Arbiter are required to be timezone localized - that seems like a hassle though.

This sounds right. And a complicating factor for SRML was nighttime 0s in the future if we requested a day from the current month.

Reply:
Here's a [raw data file](http://solardat.uoregon.edu/download/Archive/CYRO2306.txt) for the current month that shows -999 and similar for future dates.

![image](https://github.com/pvlib/pvlib-python/assets/4383303/39aa5bdd-8cf2-48fe-90e9-9d19338d8080)

Here's a [raw data file](http://solardat.uoregon.edu/download/Archive/STPO2306.txt) that includes -999 and similar as well as 0s

![image](https://github.com/pvlib/pvlib-python/assets/4383303/779691e6-6461-4f7b-96f2-275daeb7748f)

This caused a problem in SFA https://github.com/SolarArbiter/solarforecastarbiter-core/pull/572 and it's reasonable to expect that it would cause a problem with other user code. I don't know if that's pvlib's problem to solve, but I think it's somewhat more likely to come up with this new function that accepts datetimes instead of entire months.

Reply:
I don't see how the new function makes this more of an issue, currently, I would still request the same months with the old function, but it would just be more manual work.

I do think that it deserves a Warning entry and perhaps we can also implement a line that cuts off future data? For example:

    data = data.loc[:pd.Timestamp.today(), :]

Reply:
I think many users would have different expectations of the new function that accepts start, end datetimes than the old function that accepts a year and a month. 

Thanks for adding the warning. I'd rather see the previously rejected start:end slicing than .today() slicing. I'm also fine with just adding the warning and seeing if users complain.